src/ore/src/metrics.rs

```diff
@@ -403,5 +408,481 @@ pub type ComputedIntGauge = ComputedGenericGauge<AtomicI64>;
+        ExecTimeFuture {
+            fut: self,
+            metric: UnspecifiedMetric(()),
+            running_duration: Duration::from_millis(0),
+            filter: None,
+        }
+    }
+}
+
+/// Future returned by [`MetricsFutureExt::wall_time`].
+#[must_use = "futures do nothing unless you `.await` or poll them"]
+#[pin_project]
+pub struct WallTimeFuture<F, Metric> {
+    /// The inner [`Future`] that we're recording the wall time for.
+    #[pin]
+    fut: F,
+    /// Prometheus metric that we'll report to.
+    metric: Metric,
+    /// [`Instant`] at which the [`Future`] was first polled.
+    start: Option<Instant>,
+    /// Optional filter that determines if we observe the wall time of this [`Future`].
+    filter: Option<Box<dyn FnMut(Duration) -> bool>>,
+}
+
+impl<F: Debug, M: Debug> fmt::Debug for WallTimeFuture<F, M> {
+    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
+        f.debug_struct("WallTimeFuture")
+            .field("fut", &self.fut)
+            .field("metric", &self.metric)
+            .field("start", &self.start)
+            .field("filter", &self.filter.is_some())
+            .finish()
+    }
+}
+
+impl<F> WallTimeFuture<F, UnspecifiedMetric> {
+    /// Sets the recored metric to be a [`prometheus::Histogram`].
+    ///
+    /// ```text
+    /// my_future
+    ///     .wall_time()
+    ///     .observe(metrics.slow_queries_hist.with_label_values(&["select"]))
+    /// ```
+    pub fn observe(
+        self,
+        histogram: prometheus::Histogram,
+    ) -> WallTimeFuture<F, prometheus::Histogram> {
+        WallTimeFuture {
+            fut: self.fut,
+            metric: histogram,
+            start: self.start,
+            filter: self.filter,
+        }
+    }
+
+    /// Sets the recored metric to be a [`prometheus::Counter`].
+    ///
+    /// ```text
+    /// my_future
+    ///     .wall_time()
+    ///     .inc_by(metrics.slow_queries.with_label_values(&["select"]))
+    /// ```
+    pub fn inc_by(self, counter: prometheus::Counter) -> WallTimeFuture<F, prometheus::Counter> {
+        WallTimeFuture {
+            fut: self.fut,
+            metric: counter,
+            start: self.start,
+            filter: self.filter,
+        }
+    }
+}
+
+impl<F, M> WallTimeFuture<F, M> {
+    /// Specifies a filter which much return `true` for the wall time to be recorded.
+    ///
+    /// This can be particularly useful if you have a high volume `Future`, e.g. Coordinator
+    /// messages, and you only want to record ones that take a long time to complete.
+    pub fn with_filter(mut self, filter: impl FnMut(Duration) -> bool + 'static) -> Self {
+        self.filter = Some(Box::new(filter));
+        self
+    }
+}
+
+impl<F: Future, M: DurationMetric> Future for WallTimeFuture<F, M> {
+    type Output = F::Output;
+
+    fn poll(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+        let this = self.project();
+
+        // Record when we started this Future.
```

Comment:
nit: I think the inline comments here are self-explanatory with the code :)

Reply:
Done!